docs/source/how_to_guides/cb_estimators.ipynb

```diff
@@ -0,0 +1,144 @@
+{
+ "cells": [
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Using VW-Estimators\n",
+    "\n",
+    "`vw-estimators` is a library of off-policy estimators for various problems including contextual bandits. They can be used to evaluate target policies against a logged contextual bandit dataset. This library includes confidence bounds in addition to the estimators. In this example we process a trivial example dataset and feed the results into an IPS estimator and CressieRead confidence interval.\n",
+    "\n",
+    "`extract_label` is a function to translate how VW represents the contextual bandit label information into a more familiar form."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from typing import List, Optional, Tuple\n",
+    "import vowpal_wabbit_next as vw\n",
+    "from estimators.bandits import ips, cressieread\n",
+    "\n",
+    "# VW's labels contain extra info, and are associated with each example.\n",
+    "# This function extracts the logical CB label from the example list.\n",
+    "# Assumes examples have CBLabel typed labels.\n",
+    "def extract_label(examples: List[vw.Example]) -> Optional[Tuple[int, float, float]]:\n",
+    "    first_is_shared = len(examples) > 0 and examples[0].get_label().shared\n",
+    "    for i, example in enumerate(examples):\n",
+    "        if (label := example.get_label().label) is not None:\n",
+    "            _, cost, prob = label\n",
+    "            return (i - (1 if first_is_shared else 0), cost, prob)\n",
+    "    return None"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "See comments for an explanation of the process."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "[warning] Any duplicate namespace interactions will be removed\n",
+      "You can use --leave_duplicate_interactions to disable this behaviour.\n",
+      "\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Estimate: -0.5000000153668225\n",
+      "Lower bound: -1.5\n",
+      "Upper bound: 2.0\n"
+     ]
+    }
+   ],
+   "source": [
+    "workspace = vw.Workspace([\"--cb_explore_adf\", \"-q::\"])\n",
+    "estimator = ips.Estimator()\n",
+    "interval = cressieread.Interval(empirical_r_bounds=True)\n",
+    "\n",
+    "estimates = []\n",
+    "lower = []\n",
+    "upper = []\n",
+    "\n",
+    "with open(\"example_cb.txt\") as dataset_file:\n",
+    "    with vw.TextFormatReader(workspace, dataset_file) as reader:\n",
+    "        for example in reader:\n",
```

Comment:
"example" here vs "examples" in the next line (extract_label signature) is a bit confusing

Reply:
True. Perhaps I should rename to event?

Reply:
yeah I think it is better